src/parquet/file/writer-internal.cc

```diff
@@ -133,50 +133,38 @@ int64_t SerializedPageWriter::WriteDictionaryPage(const DictionaryPage& page) {
 // RowGroupSerializer
 
 int RowGroupSerializer::num_columns() const {
-  return schema_->num_columns();
+  return metadata_->num_columns();
 }
 
 int64_t RowGroupSerializer::num_rows() const {
   return num_rows_;
 }
 
-const SchemaDescriptor* RowGroupSerializer::schema() const {
-  return schema_;
-}
-
 ColumnWriter* RowGroupSerializer::NextColumn() {
-  if (current_column_index_ == schema_->num_columns() - 1) {
-    throw ParquetException("All columns have already been written.");
-  }
-  current_column_index_++;
-
-  if (current_column_writer_) { total_bytes_written_ += current_column_writer_->Close(); }
-
-  const ColumnDescriptor* column_descr = schema_->Column(current_column_index_);
-  format::ColumnChunk* col_meta = &metadata_->columns[current_column_index_];
-  col_meta->__isset.meta_data = true;
-  col_meta->meta_data.__set_type(ToThrift(column_descr->physical_type()));
-  col_meta->meta_data.__set_path_in_schema(column_descr->path()->ToDotVector());
-  std::unique_ptr<PageWriter> pager(new SerializedPageWriter(
-      sink_, properties_->compression(column_descr->path()), col_meta, allocator_));
-  current_column_writer_ =
-      ColumnWriter::Make(column_descr, std::move(pager), num_rows_, properties_);
+  // Throws an error if more columns are being written
+  auto col_meta = metadata_->NextColumnChunk();
+
+  if (current_column_writer_) { total_bytes_written_ = current_column_writer_->Close(); }
+
+  const ColumnDescriptor* column_descr = col_meta->descriptor();
+  std::unique_ptr<PageWriter> pager(
+      new SerializedPageWriter(sink_, properties_->compression(column_descr->path()),
+          col_meta, properties_->allocator()));
+  current_column_writer_ = ColumnWriter::Make(
+      col_meta->descriptor(), std::move(pager), num_rows_, properties_);
   return current_column_writer_.get();
 }
 
 void RowGroupSerializer::Close() {
   if (!closed_) {
     closed_ = true;
-    if (current_column_index_ != schema_->num_columns() - 1) {
-      throw ParquetException("Not all column were written in the current rowgroup.");
-    }
 
     if (current_column_writer_) {
-      total_bytes_written_ += current_column_writer_->Close();
+      total_bytes_written_ = current_column_writer_->Close();
       current_column_writer_.reset();
     }
-
-    metadata_->__set_total_byte_size(total_bytes_written_);
+    // Ensures all columns have been written
+    metadata_->Finish(total_bytes_written_);
```

Comment:
`total_bytes_written_` value updated by the `ColumnWriter` is actually redundant. `total_bytes_written_` is computed in the metadata builder by adding the `compressed_size` of all columns. 
I am still passing the value here for checking correctness. There is a DCHECK inside `RowGroupMetaDataBuilder::Finish()`.